File: java/401.binary-watch.java
/*
 * @lc app=leetcode id=401 lang=java
 *
 * [401] Binary Watch
 */

// @lc code=start
class Solution {

    class Time {
        final int hour;
        final int minute;
        public Time(int hour, int minute) {
            this.hour = hour;
            this.minute = minute;
        }
        public Time add(Time o) {
            return new Time(hour+o.hour, minute+o.minute);
        }
        public String toString() {
            return hour + ":" + (minute < 10 ? "0" : "") + minute;
        }
    }

    List<String> res = new LinkedList<>();

    public List<String> readBinaryWatch(int num) {
        Time[] times = new Time[] { 
            new Time(0,1), new Time(0,2), new Time(0,4), new Time(0,8),
            new Time(0,16), new Time(0,32),
            new Time(1,0), new Time(2,0), new Time(4,0), new Time(8,0)};
        if (num >= 0) backtrack(num, 0, times, new Time(0,0));
        return res;
    }

    private void backtrack(int num, int sp, Time[] times, Time time) {
        if (num == 0) {
            if (time.hour < 12 && time.minute < 60) res.add(time.toString());
            return;
        }
        for (int i = sp; times.length-i >= num; ++i) {
            backtrack(num-1, i+1, times, time.add(times[i]));
        }
    }
}
// @lc code=end

File: go/401.binary-watch.go
/*
 * @lc app=leetcode id=401 lang=golang
 *
 * [401] Binary Watch
 */

// @lc code=start
func readBinaryWatch(num int) []string {
	led := []int{1,2,4,8,1,2,4,8,16,32}
	res := []string{}
	var backtrack func(int, int, []int)
	backtrack = func(idx, n int, time []int) {
		if n == 0 {
			t1, t2 := strconv.Itoa(time[0]), strconv.Itoa(time[1])
			if len(t2) == 1 {
				t2 = "0" + t2
			}
			res = append(res, t1+":"+t2)
			return
		}
		for i := idx; i < len(led); i++ {
			if i < 4 {
				if led[i] + time[0] >= 12 {
					continue
				}
				time[0] += led[i]
				backtrack(i+1, n-1, time)
				time[0] -= led[i]
			} else {
				if led[i] + time[1] >= 60 {
					break
				}
				time[1] += led[i]
				backtrack(i+1, n-1, time)
				time[1] -= led[i]
			}
		}
	}

	backtrack(0, num, []int{0,0})
	return res
}
// @lc code=end

